src/client.rs

```diff
@@ -11,22 +11,23 @@ use crate::request::Request;
 use crate::response::Response;
 
 pub struct Client {
+    pub server_address: SocketAddr,
     handle: ConnectionHandle,
 }
 
 impl Client {
     pub fn connect(address: SocketAddr) -> impl Future<Item = Client, Error = io::Error> {
         TcpStream::connect(&address).and_then(|tcp_stream| {
             let mut executor = DefaultExecutor::current();
+            let server_address = tcp_stream.peer_addr().unwrap();
```

Comment:
Should just be able to move `address` into the the `Client` instance since it will be the same. Also removes an unwrap which is always good imo.

Reply:
I tried that, but got a borrow checker error that I still can't figure out.  

```
error[E0597]: `server_address` does not live long enough
  --> src/client.rs:29:33
   |
19 |     pub fn connect(server_address: SocketAddr) -> impl Future<Item = Client, Error = io::Error> {
   |                                                   --------------------------------------------- opaque type requires that `server_address` is borrowed for `'static`
20 |         TcpStream::connect(&server_address).and_then(|tcp_stream| {
   |                                                      ------------ value captured here
...
29 |             Ok(Client { handle, server_address })
   |                                 ^^^^^^^^^^^^^^ borrowed value does not live long enough
30 |         })
31 |     }
   |     - `server_address` dropped here while still borrowed
```

 I thought that since server_address originated from the very beginning of the program, before Client is instantiated, then it should be ok, but I guess `server_address` _could_ be freed before `client` (even though I think it isn't in this case).

I'll keep reading.  Just leaving this here in case you might have time to give me a hint (or point me to some good reading on the topic).

Reply:
It should work if you add `move` before the lambda: `.and_then(move |tcp_stream| { ... })`